packages/utils/src/fhir-deduplicate.ts

```diff
@@ -0,0 +1,417 @@
+ */
+function deduplicateConditions(conditions: Condition[]): Condition[] {
+  const sortedByCodes = sortByCodeLengths(conditions);
+  const groupedConditionsMap = groupSameConditions(sortedByCodes);
+  return combineSameConditions(groupedConditionsMap);
+}
+
+function sortByCodeLengths(conditions: Condition[]): Condition[] {
+  return conditions.sort((a, b) => {
+    const lengthA = a.code?.coding?.length || 0;
+    const lengthB = b.code?.coding?.length || 0;
+
+    if (lengthA > lengthB) return -1;
+    if (lengthA < lengthB) return 1;
+    return 0;
+  });
+}
+
+function groupSameConditions(conditions: Condition[]): Map<CompositeKey, Condition[]> {
+  const conditionMap = new Map<CompositeKey, Condition[]>();
+  for (const condition of conditions) {
+    const newKey = createCompositeKey(condition);
+
+    const existingKey = getExistingKey(conditionMap, newKey);
+    if (!existingKey) {
+      conditionMap.set(newKey, [condition]);
+    } else {
+      const mappedConditions = conditionMap.get(existingKey);
+      if (mappedConditions) conditionMap.set(existingKey, [...mappedConditions, condition]);
+    }
+  }
+  return conditionMap;
+}
+
+function createCompositeKey(condition: Condition): CompositeKey {
+  const codeKey = createKeyFromCodes(condition);
+  const dateKey = createDateKey(condition);
+
+  return {
+    code: codeKey,
+    date: dateKey,
+  };
+}
+
+const relevantCodeSystems = [SNOMED_CODE, ICD_10_CODE, ICD_9_CODE, LOINC_CODE, CPT_CODE, IMO_CODE];
+
+function createKeyFromCodes(condition: Condition): string {
+  const concept = condition.code;
+  let codeKey = "";
+  if (concept && concept.coding) {
+    for (const coding of concept.coding) {
+      const system = coding.system?.trim().toLowerCase();
+      const code = coding.code?.trim().toLowerCase();
+      if (system && code) {
+        // TODO: Decide whether other codes should also be included
+        for (const codeSystem of relevantCodeSystems) {
+          if (system.includes(codeSystem)) codeKey += code + "/";
+        }
+      }
+    }
+  }
+  if (codeKey.length === 0) {
+    const msg = "Specified codes not found for Condition";
+    console.error(`${msg}: ${condition.id}`);
+    // TODO: Preserve all conditions where the code key is empty
+  }
+  return codeKey;
+}
+
+function createDateKey(condition: Condition): string {
+  if (condition.onsetPeriod?.start) {
+    return getDateFromString(condition.onsetPeriod?.start);
+  }
+
+  // TODO: See what other condition attributes we encounter and account for those. For now, keeping the error for vis
+  const msg = "Start time not found in onsetPeriod. Check other attributes of Condition";
+  console.log(`${msg}: ${condition.id}`);
+  throw new Error(msg);
+}
+
+function getDateFromString(dateString: string): string {
+  const date = new Date(dateString);
+  return date.toLocaleDateString();
+}
+
+function getExistingKey<T>(
```

Comment:
the idea of using keys and a map (which is basically a hash table) would be that you would expect to be performing O(1) lookups, where you just check if the key is in the map to see if it exists. 
What we are doing here is O(n*k**2) where n is the number of keys in the map (iterating through all of them with the for loop) and k is the number of codes in each key (iterating through splitKey with "some" and iterating through splitMapKey with "includes")

This is a really innefficient approach. 

A much simpler approach would just be to make a key which is just one code at a time. 

So if a condition has 4 codes in it, for each code do 

key = concat(code+codesystem+date) 
value = Condition

and then lookups are O(1)

Reply:
I see why you took the approach you did. Your trying to combine resources across different code systems, but imo its a bit too hacky.

The simple simple thing to do is just have a simple key on the SNOMED code of the condition. If the condition doesn't have a snomed code but does have another code, do a crosswalk lookup in the term server for it. Every time we get a hit in the hash table for a condition thats already there, we could merge the two resources if we wanted / just add the removed resource to the paper trail of the master resource.
(not sure how we want to handle dates. Is it valuable to have otherwise duplicated conditions, but one is just older than the other?) 

I think as a place to start, even without the term server, a simple snomed lookup to dedup makes sense. SNOMED is the most common code system we get for conditions.

Reply:
its simple and will cover a majority of the conditions duplicates while we figure out the term server and get whatever is left.

Reply:
Thanks for the feedback!

I ran some tests on a couple files, and it seems around ~94% conditions have snomed codes:
```
file 1: 322/352
file 2: 122/127
file 3: 163/166
```

So good call on that! 
And I think it's a good idea that this first iteration should be ok to be done without a term server. We can just add that step when the term server is ready.

As for the dates, I think it makes sense to keep separate instances of a condition apart. Like, it would be valuable for a doctor to know that someone got covid 3 different times, for example (all else being equal). But I think there should be a threshold for how far apart is considered a separate instance. Because I've seen some `Condition` resources which are for the same disease but show the onset one day apart. And imo, you can't get sick twice with the same disease within a week - that would just be considered the same instance. But maybe I'm wrong 🤷 Maybe things like food poisoning are fully realistic to get twice within the same week.

Reply:
More discussion for the approach happened on Slack - [link](https://metriport.slack.com/archives/C04DMKE9DME/p1722443874726129)